src/ol/render/canvas/ZIndexContext.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * @module ol/render/canvas/ZIndexContext
+ */
+
+/** @typedef {CanvasRenderingContext2D & {globalAlpha: any}} ZIndexContextProxy */
+
+/**
+ * @extends {CanvasRenderingContext2D}
+ */
+class ZIndexContext {
+  constructor() {
+    /**
+     * @private
+     * @type {Array<Array<*>>}
+     */
+    this.instructions_ = [];
+    /**
+     * @type {number}
+     */
+    this.zIndex = 0;
+    /**
+     * @private
+     * @type {number}
+     */
+    this.offset_ = 0;
+
+    /**
+     * @type {ZIndexContextProxy}
+     */
+    this.context_ = /** @type {ZIndexContextProxy} */ (
+      new Proxy(CanvasRenderingContext2D.prototype, {
+        get: (target, property) => {
+          if (property === 'globalAlpha' || property === 'canvas') {
+            return undefined;
+          }
```

Comment:
As described above, the proxy returns `undefined` for state properties, because the recorded instructions do not know about the state of the target context.

Reply:
This is cool.  Maybe this is unnecessary if we only try to access `globalAlpha` and `canvas` properties, but it seems like a more general guard might check the type of the property on the target.
```js
if (typeof target[property] !== 'function') {
    // we only accept calling functions on the proxy, not accessing properties
    return undefined;
}
```

Reply:
Agreed it makes sense to return undefined for all non-function properties. What I do below is the opposite case (functions). Pushed a saparate commit with this simplification.